docs/features/routing.rst

```diff
@@ -235,4 +235,36 @@ Ocelot will also allow you to put query string parameters in the UpstreamPathTem
     }
 
 In this example Ocelot will only match requests that have a matching url path and the query string starts with unitId=something. You can have other queries after this
-but you must start with the matching parameter. Also Ocelot will swap the {unitId} parameter from the query string and use it in the downstream request path. 
+but you must start with the matching parameter. Also Ocelot will swap the {unitId} parameter from the query string and use it in the downstream request path.
+
+Upstream header-based routing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This feature was requested in `issue 360 <https://github.com/ThreeMammals/Ocelot/issues/360>`_ and `issue 624 <https://github.com/ThreeMammals/Ocelot/issues/624>`_.
+
+Ocelot allows you to define a Route with upstream headers, each of which may define a set of accepted values. If a Route has a set of upstream headers defined in it, it will no longer match a request's upstream path based solely on upstream path template. The request must also contain one or more headers required by the Route for a match.
+
+A sample configuration might look like the following:
+
+.. code-block:: json
+
+    {
+        "Routes": [
+            {
+                "UpstreamHeaderRoutingOptions": {
+                    "Headers": {
+                        "X-API-Version": [ "1", "2" ],
+                        "X-Tennant-Id": [ "tennantId" ]
+                    },
+                    "TriggerOn": "all"
+                }
+            }
+        ]
+    }
+
+The ``UpstreamHeaderRoutingOptions`` block defines two attributes -- the ``Headers`` block and the ``TriggerOn`` attribute. The ``Headers`` attribute defines required header names as keys and lists of acceptable header values as values. During route matching, both header names and values are matched in *case insensitive* manner. Please note that if a header has more than one acceptable value configured, presence of any of those values in a request is sufficient for a header to be a match.
+
+The second attribute, ``TriggerOn``, defines how the route finder will determine whether a particular header configuration in a request matches a Route's header configuration. The attribute accepts two values:
+
+* ``"Any"`` causes the route finder to match a Route if any value of any configured header is present in a request
+* ``"All"`` causes the route finder to match a Route only if any value of *all* configured headers is present in a request

```

Comment:
No, it doesn't matter as far as parsing. See `UpstreamHeaderRoutingOptionsCreator`, it's parsed in a case-insensitive manner. Any unit test should reflect, otherwise it will be too brittle.

Reply:
Sure, I can search for this class and review once again.
Please, share a link to exact code line in your feature branch.

Reply:
Here you go: https://github.com/mrclayman/Ocelot/blob/4bc297aa527cb3bd638c0b683fdd4ed673d37786/src/Ocelot/Configuration/Creator/UpstreamHeaderRoutingOptionsCreator.cs#L15

Reply:
🆗 Not an issue